File: backend/src/security.ts
import { createHmac, pbkdf2Sync, randomBytes, timingSafeEqual } from "node:crypto";

export type TokenPayload = {
  sub: string;
  email: string;
  role: "admin" | "user";
  exp: number;
};

const HASH_ALGORITHM = "sha256";
const HASH_ITERATIONS = 80_000;
const HASH_BYTES = 32;

export function hashPassword(password: string) {
  const salt = randomBytes(18).toString("base64url");
  const hash = pbkdf2Sync(password, salt, HASH_ITERATIONS, HASH_BYTES, HASH_ALGORITHM).toString("base64url");
  return `pbkdf2$${HASH_ITERATIONS}$${salt}$${hash}`;
}

export function verifyPassword(password: string, verifier: string) {
  const [scheme, iterationsText, salt, expectedHash] = verifier.split("$");
  if (scheme !== "pbkdf2" || !iterationsText || !salt || !expectedHash) return false;

  const iterations = Number(iterationsText);
  if (!Number.isInteger(iterations) || iterations < 1) return false;

  const actual = Buffer.from(pbkdf2Sync(password, salt, iterations, HASH_BYTES, HASH_ALGORITHM).toString("base64url"));
  const expected = Buffer.from(expectedHash);
  return actual.length === expected.length && timingSafeEqual(actual, expected);
}

export function signToken(payload: Omit<TokenPayload, "exp">, secret: string, ttlSeconds = 60 * 60 * 24) {
  const body: TokenPayload = {
    ...payload,
    exp: Math.floor(Date.now() / 1000) + ttlSeconds
  };
  const encoded = base64url(JSON.stringify(body));
  const signature = createHmac("sha256", secret).update(encoded).digest("base64url");
  return `${encoded}.${signature}`;
}

export function verifyToken(token: string, secret: string): TokenPayload | null {
  const [encoded, signature] = token.split(".");
  if (!encoded || !signature) return null;

  const expected = createHmac("sha256", secret).update(encoded).digest("base64url");
  if (!safeEqual(signature, expected)) return null;

  try {
    const payload = JSON.parse(Buffer.from(encoded, "base64url").toString("utf8")) as TokenPayload;
    if (!payload.sub || !payload.email || !payload.role || payload.exp < Math.floor(Date.now() / 1000)) return null;
    if (payload.role !== "admin" && payload.role !== "user") return null;
    return payload;
  } catch {
    return null;
  }
}

function base64url(value: string) {
  return Buffer.from(value).toString("base64url");
}

function safeEqual(left: string, right: string) {
  const leftBuffer = Buffer.from(left);
  const rightBuffer = Buffer.from(right);
  return leftBuffer.length === rightBuffer.length && timingSafeEqual(leftBuffer, rightBuffer);
}

File: apps/api/app/security.py
import base64
import hashlib
import hmac
import json
import secrets
import time
from typing import Any

from app.core.config import get_settings


def hash_password(password: str) -> str:
    salt = secrets.token_urlsafe(18)
    iterations = 120_000
    digest = hashlib.pbkdf2_hmac("sha256", password.encode("utf-8"), salt.encode("utf-8"), iterations, dklen=32)
    return f"pbkdf2${iterations}${salt}${base64.urlsafe_b64encode(digest).decode('utf-8')}"


def verify_password(password: str, verifier: str) -> bool:
    try:
        scheme, iterations_text, salt, expected = verifier.split("$", 3)
        if scheme != "pbkdf2":
            return False
        iterations = int(iterations_text)
        actual = hashlib.pbkdf2_hmac("sha256", password.encode("utf-8"), salt.encode("utf-8"), iterations, dklen=32)
        return hmac.compare_digest(base64.urlsafe_b64encode(actual).decode("utf-8"), expected)
    except (ValueError, TypeError):
        return False


def _b64(value: bytes) -> str:
    return base64.urlsafe_b64encode(value).decode("utf-8").rstrip("=")


def _unb64(value: str) -> bytes:
    padded = value + "=" * (-len(value) % 4)
    return base64.urlsafe_b64decode(padded.encode("utf-8"))


def sign_token(payload: dict[str, Any], ttl_seconds: int | None = None) -> str:
    settings = get_settings()
    expires_in = ttl_seconds or settings.access_token_expire_minutes * 60
    body = {**payload, "exp": int(time.time()) + expires_in}
    encoded = _b64(json.dumps(body, separators=(",", ":"), ensure_ascii=False).encode("utf-8"))
    signature = hmac.new(settings.app_secret_key.encode("utf-8"), encoded.encode("utf-8"), hashlib.sha256).digest()
    return f"{encoded}.{_b64(signature)}"


def verify_token(token: str) -> dict[str, Any] | None:
    settings = get_settings()
    try:
        encoded, signature = token.split(".", 1)
        expected = _b64(hmac.new(settings.app_secret_key.encode("utf-8"), encoded.encode("utf-8"), hashlib.sha256).digest())
        if not hmac.compare_digest(signature, expected):
            return None
        payload = json.loads(_unb64(encoded).decode("utf-8"))
        if int(payload.get("exp", 0)) < int(time.time()):
            return None
        return payload
    except (ValueError, TypeError, json.JSONDecodeError):
        return None
